app/utils/SQL.scala

```diff
@@ -0,0 +1,251 @@
+package utils
+
+import com.scalableminds.util.time.Instant
+import play.api.libs.json.{JsValue, Json}
+import slick.dbio.Effect
+import slick.jdbc._
+import slick.sql.SqlStreamingAction
+import slick.util.DumpInfo
+
+import java.sql.{PreparedStatement, Types}
+import scala.collection.mutable
+import scala.collection.mutable.ListBuffer
+import scala.concurrent.duration
+import scala.concurrent.duration.Duration
+
+class NestedSQLInterpolator(val s: StringContext) extends AnyVal {
+  def nsql(param: Any*): SqlToken = {
+    val parts = s.parts.toList
+    val values = param.toList
+
+    val outputSql = mutable.StringBuilder.newBuilder
+    val outputValues = ListBuffer[SqlValue]()
+
+    assert(parts.length == values.length + 1)
+    for (i <- parts.indices) {
+      outputSql ++= parts(i);
+
+      if (i < values.length) {
+        val value = values(i)
+        value match {
+          case x: SqlToken => {
+            outputSql ++= x.sql
+            outputValues ++= x.values
+          }
+          case x => {
+            val sqlValue = SqlValue.makeSqlValue(x)
+            outputSql ++= sqlValue.getPlaceholder()
+            outputValues += sqlValue
+          }
+        }
+      }
+    }
+
+    SqlToken(sql = outputSql.toString, values = outputValues.toList)
+  }
+}
+
+object NestedSQLInterpolation {
+  implicit def nestedSQLInterpolation(s: StringContext): NestedSQLInterpolator = new NestedSQLInterpolator(s)
+}
+
+case class SqlToken(sql: String, values: List[SqlValue] = List()) {
+  def debug: String = {
+    val parts = sql.split("\\?", -1)
+    assert(parts.tail.length == values.length)
+    parts.tail.zip(values).foldLeft(parts.head)((acc, x) => acc + x._2.debug() + x._1)
+  }
+
+  def as[R](implicit rconv: GetResult[R]): SqlStreamingAction[Vector[R], R, Effect] =
+    new StreamingInvokerAction[Vector[R], R, Effect] {
+      def statements = List(sql)
+
+      protected[this] def createInvoker(statements: Iterable[String]) = new StatementInvoker[R] {
+        val getStatement = statements.head
+
+        protected def setParam(st: PreparedStatement) = {
+          val pp = new PositionedParameters(st);
+          values.foreach(_.setParameter(pp))
+        }
+
+        protected def extractValue(rs: PositionedResult): R = rconv(rs)
+      }
+
+      override def getDumpInfo = DumpInfo(DumpInfo.simpleNameFor(getClass), mainInfo = s"[${debug}]")
+
+      protected[this] def createBuilder = Vector.newBuilder[R]
+    }
+
+  def asUpdate = as[Int](GetUpdateValue).head
+
+  def stripMargin(marginChar: Char): SqlToken = copy(sql = sql.stripMargin(marginChar))
+  def stripMargin: SqlToken = copy(sql = sql.stripMargin)
```

Comment:
those seem unused? are they supposed to be part of the API?

Reply:
I had them in, because they are common with `s` and `sql[u]`-based string interpolation. I took them out again, because they don't seem to be used anywhere in our codebase.